{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3627005f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "y\n",
      "y\n",
      "y\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "yes | head -3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b7294677",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqKit -- a cross-platform and ultrafast toolkit for FASTA/Q file manipulation\n",
      "\n",
      "Version: 2.1.0\n",
      "\n",
      "Author: Wei Shen <shenwei356@gmail.com>\n",
      "\n",
      "Documents  : http://bioinf.shenwei.me/seqkit\n",
      "Source code: https://github.com/shenwei356/seqkit\n",
      "Please cite: https://doi.org/10.1371/journal.pone.0163962\n",
      "\n",
      "\n",
      "Seqkit utlizies the pgzip (https://github.com/klauspost/pgzip) package to\n",
      "read and write gzip file, and the outputted gzip file would be slighty\n",
      "larger than files generated by GNU gzip.\n",
      "\n",
      "Seqkit writes gzip files very fast, much faster than the multi-threaded pigz,\n",
      "therefore there's no need to pipe the result to gzip/pigz.\n",
      "\n",
      "Usage:\n",
      "  seqkit [command]\n",
      "\n",
      "Available Commands:\n",
      "  amplicon        extract amplicon (or specific region around it) via primer(s)\n",
      "  bam             monitoring and online histograms of BAM record features\n",
      "  common          find common sequences of multiple files by id/name/sequence\n",
      "  concat          concatenate sequences with same ID from multiple files\n",
      "  convert         convert FASTQ quality encoding between Sanger, Solexa and Illumina\n",
      "  duplicate       duplicate sequences N times\n",
      "  faidx           create FASTA index file and extract subsequence\n",
      "  fish            look for short sequences in larger sequences using local alignment\n",
      "  fq2fa           convert FASTQ to FASTA\n",
      "  fx2tab          convert FASTA/Q to tabular format (and length, GC content, average quality...)\n",
      "  genautocomplete generate shell autocompletion script (bash|zsh|fish|powershell)\n",
      "  grep            search sequences by ID/name/sequence/sequence motifs, mismatch allowed\n",
      "  head            print first N FASTA/Q records\n",
      "  head-genome     print sequences of the first genome with common prefixes in name\n",
      "  locate          locate subsequences/motifs, mismatch allowed\n",
      "  mutate          edit sequence (point mutation, insertion, deletion)\n",
      "  pair            match up paired-end reads from two fastq files\n",
      "  range           print FASTA/Q records in a range (start:end)\n",
      "  rename          rename duplicated IDs\n",
      "  replace         replace name/sequence by regular expression\n",
      "  restart         reset start position for circular genome\n",
      "  rmdup           remove duplicated sequences by ID/name/sequence\n",
      "  sample          sample sequences by number or proportion\n",
      "  sana            sanitize broken single line FASTQ files\n",
      "  scat            real time recursive concatenation and streaming of fastx files\n",
      "  seq             transform sequences (extract ID, filter by length, remove gaps...)\n",
      "  shuffle         shuffle sequences\n",
      "  sliding         extract subsequences in sliding windows\n",
      "  sort            sort sequences by id/name/sequence/length\n",
      "  split           split sequences into files by id/seq region/size/parts (mainly for FASTA)\n",
      "  split2          split sequences into files by size/parts (FASTA, PE/SE FASTQ)\n",
      "  stats           simple statistics of FASTA/Q files\n",
      "  subseq          get subsequences by region/gtf/bed, including flanking sequences\n",
      "  tab2fx          convert tabular format to FASTA/Q format\n",
      "  translate       translate DNA/RNA to protein sequence (supporting ambiguous bases)\n",
      "  version         print version information and check for update\n",
      "  watch           monitoring and online histograms of sequence features\n",
      "\n",
      "Flags:\n",
      "      --alphabet-guess-seq-length int   length of sequence prefix of the first FASTA record based on which seqkit guesses the sequence type (0 for whole seq) (default 10000)\n",
      "  -h, --help                            help for seqkit\n",
      "      --id-ncbi                         FASTA head is NCBI-style, e.g. >gi|110645304|ref|NC_002516.2| Pseud...\n",
      "      --id-regexp string                regular expression for parsing ID (default \"^(\\\\S+)\\\\s?\")\n",
      "      --infile-list string              file of input files list (one file per line), if given, they are appended to files from cli arguments\n",
      "  -w, --line-width int                  line width when outputting FASTA format (0 for no wrap) (default 60)\n",
      "  -o, --out-file string                 out file (\"-\" for stdout, suffix .gz for gzipped out) (default \"-\")\n",
      "      --quiet                           be quiet and do not show extra information\n",
      "  -t, --seq-type string                 sequence type (dna|rna|protein|unlimit|auto) (for auto, it automatically detect by the first sequence) (default \"auto\")\n",
      "  -j, --threads int                     number of CPUs. can also set with environment variable SEQKIT_THREADS) (default 4)\n",
      "\n",
      "Use \"seqkit [command] --help\" for more information about a command.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "seqkit --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "595dbce4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
