{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "960dfa4a-dc11-4f8d-a131-4767bc59b90b",
   "metadata": {},
   "outputs": [],
   "source": [
    "IDX=refs/genome.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "03f642cf-1c04-4cbd-b48a-13f0926f3ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"refs/genome.fa.*.ht2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Local offset rate: 3 (one in 8)\n",
      "  Local fTable chars: 6\n",
      "  Local sequence length: 57344\n",
      "  Local sequence overlap between two consecutive indexes: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  refs/genome.fa\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to read SNPs and splice sites: 00:00:00\n",
      "Using parameters --bmax 24141 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 24141 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 18392.7 (target: 24140)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering GFM loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (24141) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 11794 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 11795 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (24141) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 15630 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 15631 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (24141) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 19642 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 19643 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (24141) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 19091 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 19092 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (24141) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 21618 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 21619 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (24141) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 22719 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 22720 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (24141) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 18255 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 18256 for bucket 7\n",
      "Exited GFM loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 47248\n",
      "fchr[G]: 65274\n",
      "fchr[T]: 83911\n",
      "fchr[$]: 128755\n",
      "Exiting GFM::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4237485 bytes to primary GFM file: refs/genome.fa.1.ht2\n",
      "Wrote 32196 bytes to secondary GFM file: refs/genome.fa.2.ht2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from GFM constructor\n",
      "Returning from initFromVector\n",
      "Wrote 62331 bytes to primary GFM file: refs/genome.fa.5.ht2\n",
      "Wrote 32708 bytes to secondary GFM file: refs/genome.fa.6.ht2\n",
      "Re-opening _in5 and _in5 as input streams\n",
      "Returning from HGFM constructor\n",
      "Headers:\n",
      "    len: 128755\n",
      "    gbwtLen: 128756\n",
      "    nodes: 128756\n",
      "    sz: 32189\n",
      "    gbwtSz: 32190\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 0\n",
      "    eftabSz: 0\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 8048\n",
      "    offsSz: 32192\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideGbwtSz: 48\n",
      "    sideGbwtLen: 192\n",
      "    numSides: 671\n",
      "    numLines: 671\n",
      "    gbwtTotLen: 42944\n",
      "    gbwtTotSz: 42944\n",
      "    reverse: 0\n",
      "    linearFM: Yes\n",
      "Total time for call to driver() for forward index: 00:00:00\n"
     ]
    }
   ],
   "source": [
    "hisat2-build $IDX $IDX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "243aabfb-62fc-4706-a48e-ff91835f321b",
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools faidx $IDX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6ed9d7b5-9a83-4fbb-b38b-0e74f79e3f5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.0\tSO:unsorted\n",
      "@SQ\tSN:Golden\tLN:128756\n",
      "@PG\tID:hisat2\tPN:hisat2\tVN:2.2.1\tCL:\"/srv/conda/condax/envs/hisat2/bin/hisat2-align-s --wrapper basic-0 -x refs/genome.fa --read-lengths 100 -1 reads/BORED_1_R1.fq -2 reads/BORED_1_R2.fq\"\n",
      "HWI-ST718_146963544:7:1101:1307:81391\t83\tGolden\t33031\t60\t100M\t=\t32827\t-304\tACCGCCGCATACGGCCGATTGTCGCAGCCCGGGTCGATTATAACAACGGTGCAATCTCAGCTAAACCGACGCAGTTTTGCTCCTTGGATTCTGAGCCCGG\t<BB><5CA<>5B@@B@C>@<@<5?<<>?@BB@>B=C@AADCBB?56HHDGGHIHEGFGEHIGGHFHAGDIGGEIIIIIHCC;AIIIGFHHHHFFFDFCCC\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:100\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "HWI-ST718_146963544:7:1101:1307:81391\t163\tGolden\t32827\t60\t100M\t=\t33031\t304\tCATGCACAGCGGTCAAACAGTATGTCCCAAGGGGACTTAAGCGCGGTGGCCTCCCCTATCCCCTACGAGGCTACCCGGATCGATGACGCGAATTGGGGAC\t@C?DFFFFHHHH?FHIGIIIHGHIIIIIGHIGBD@FHIIIIIGGED@EBC?;BDFDEDEEDDDCDD@BDDB?CCCBB59@DBBBAC>BDB9952>CBD05\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:100\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "HWI-ST718_146963544:7:1101:1328:22995\t83\tGolden\t33653\t60\t100M\t=\t33534\t-219\tATCGCGCACGTAGTAGCCTAGAGCGCCAGGGGCGGAAATTCGCCTGAAAAGTTTTGCCGGCGCACAAGCACGATCGGCTCCTAATAGGAGGTGAATTAGA\tBBBBBDDBDDDCCCCCCCCA<-BBBABDDDDDDEDDDDDDDDEECEFFFDBEEHJJJIHEGGGFIGIHGFJIJJJJJJJIJJJJJIJHHHHHFFFFFCCC\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:100\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "HWI-ST718_146963544:7:1101:1328:22995\t163\tGolden\t33534\t60\t100M\t=\t33653\t219\tCTAAGCTCGGCAATCGCTCTTGCCGTGCGAGATGCTAGCAAAAAGGTGTACTTCTCAGCGGAGCAGAAAGATCATGTTTATTGGAAGCATCAACCTGCGC\tBB@FFFFFGHHHGJJJJJJJJFJIIFHHHGIGEHIIJIJIJJIGGI7CEHCHIHGHHHHFF6@BDDCCDCDDDDCDCDDDDDDDDDCBDCDDDDDCDDD>\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:100\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "HWI-ST718_146963544:7:1101:1378:31780\t83\tGolden\t616\t60\t100M\t=\t556\t-160\tCCTCCTGTTTGCCACGCTGGACCCGATGACCAGAAAAATGGTCCTGCCAAGCGGCTACAGTGTTCTTCTTTCAGATACAGTAGGATTTATTCAGGATCTT\tBBA8CDDBCCDB=<=C@3BBACDDBB=ACFFHGJJIHAJJIJJJJJIGHJIIIHIHHHJIGJJJIJJIJJJJJJJIGIJJJJJIJJJHHHHHFFFFF@CC\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:100\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "HWI-ST718_146963544:7:1101:1378:31780\t163\tGolden\t556\t60\t100M\t=\t616\t160\tAAACGCAGGGAAATCAACATGGTTCAACCGCCTGACGAGTGCTGACAGCTATGAAGAAGACCTCCTGTTTGCCACGCTGGACCCGATGACCAGAAAAATG\t@CCFFFFFHHHHHIIIJJJJJJFHIJJJJHIJGHJIIJJGHHIJJHIJJJHIJHHHHGFFFFFEDEECEDDDDDDDDDDDDDDDDDDDDDDDDDCDDBAC\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:100\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "HWI-ST718_146963544:7:1101:1402:16348\t83\tGolden\t32913\t60\t100M\t=\t32833\t-180\tCGCGAATTGGGGACATTCAAATGAGCATCCTAGTCACCGCGTTTAAAATGAACCTGCCGGCTGATCGTTTTTAGGATATTGTGAGTAATATAGACTGGCG\tDDDDDCDDDDEEDDDDEEEEDDDCADDDDDDDDDB?CDHCHJJJJJJIIIGCJIJJJJJIGIIHIIJJJJJJJJJJJIJJJJJIJJJHHHHHDDFFFCCC\tAS:i:-5\tZS:i:-6\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:94T5\tYS:i:0\tYT:Z:CP\tNH:i:1\n",
      "(ERR): hisat2-align exited with value 141\n"
     ]
    }
   ],
   "source": [
    "hisat2 -x refs/genome.fa -1 reads/BORED_1_R1.fq  -2 reads/BORED_1_R2.fq | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "931407bd-8884-4f17-895f-c4cdd8c87d34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Academic tradition requires you to cite works you base your article on.\n",
      "If you use programs that use GNU Parallel to process data for an article in a\n",
      "scientific publication, please cite:\n",
      "\n",
      "  Tange, O. (2022, September 22). GNU Parallel 20220922 ('Elizabeth').\n",
      "  Zenodo. https://doi.org/10.5281/zenodo.7105792\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "More about funding GNU Parallel and the citation notice:\n",
      "https://www.gnu.org/software/parallel/parallel_design.html#citation-notice\n",
      "\n",
      "To silence this citation notice: run 'parallel --citation' once.\n",
      "\n",
      "112193 reads; of these:\n",
      "  112193 (100.00%) were paired; of these:\n",
      "    319 (0.28%) aligned concordantly 0 times\n",
      "    111874 (99.72%) aligned concordantly exactly 1 time\n",
      "    0 (0.00%) aligned concordantly >1 times\n",
      "    ----\n",
      "    319 pairs aligned concordantly 0 times; of these:\n",
      "      316 (99.06%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6 mates make up the pairs; of these:\n",
      "        0 (0.00%) aligned 0 times\n",
      "        6 (100.00%) aligned exactly 1 time\n",
      "        0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "137581 reads; of these:\n",
      "  137581 (100.00%) were paired; of these:\n",
      "    21838 (15.87%) aligned concordantly 0 times\n",
      "    115743 (84.13%) aligned concordantly exactly 1 time\n",
      "    0 (0.00%) aligned concordantly >1 times\n",
      "    ----\n",
      "    21838 pairs aligned concordantly 0 times; of these:\n",
      "      21004 (96.18%) aligned discordantly 1 time\n",
      "    ----\n",
      "    834 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      1668 mates make up the pairs; of these:\n",
      "        0 (0.00%) aligned 0 times\n",
      "        1668 (100.00%) aligned exactly 1 time\n",
      "        0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "123093 reads; of these:\n",
      "  123093 (100.00%) were paired; of these:\n",
      "    371 (0.30%) aligned concordantly 0 times\n",
      "    122696 (99.68%) aligned concordantly exactly 1 time\n",
      "    26 (0.02%) aligned concordantly >1 times\n",
      "    ----\n",
      "    371 pairs aligned concordantly 0 times; of these:\n",
      "      363 (97.84%) aligned discordantly 1 time\n",
      "    ----\n",
      "    8 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      16 mates make up the pairs; of these:\n",
      "        0 (0.00%) aligned 0 times\n",
      "        14 (87.50%) aligned exactly 1 time\n",
      "        2 (12.50%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "237018 reads; of these:\n",
      "  237018 (100.00%) were paired; of these:\n",
      "    375 (0.16%) aligned concordantly 0 times\n",
      "    236643 (99.84%) aligned concordantly exactly 1 time\n",
      "    0 (0.00%) aligned concordantly >1 times\n",
      "    ----\n",
      "    375 pairs aligned concordantly 0 times; of these:\n",
      "      365 (97.33%) aligned discordantly 1 time\n",
      "    ----\n",
      "    10 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      20 mates make up the pairs; of these:\n",
      "        0 (0.00%) aligned 0 times\n",
      "        20 (100.00%) aligned exactly 1 time\n",
      "        0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "158009 reads; of these:\n",
      "  158009 (100.00%) were paired; of these:\n",
      "    536 (0.34%) aligned concordantly 0 times\n",
      "    157473 (99.66%) aligned concordantly exactly 1 time\n",
      "    0 (0.00%) aligned concordantly >1 times\n",
      "    ----\n",
      "    536 pairs aligned concordantly 0 times; of these:\n",
      "      515 (96.08%) aligned discordantly 1 time\n",
      "    ----\n",
      "    21 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      42 mates make up the pairs; of these:\n",
      "        0 (0.00%) aligned 0 times\n",
      "        42 (100.00%) aligned exactly 1 time\n",
      "        0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "196673 reads; of these:\n",
      "  196673 (100.00%) were paired; of these:\n",
      "    660 (0.34%) aligned concordantly 0 times\n",
      "    195995 (99.66%) aligned concordantly exactly 1 time\n",
      "    18 (0.01%) aligned concordantly >1 times\n",
      "    ----\n",
      "    660 pairs aligned concordantly 0 times; of these:\n",
      "      643 (97.42%) aligned discordantly 1 time\n",
      "    ----\n",
      "    17 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      34 mates make up the pairs; of these:\n",
      "        0 (0.00%) aligned 0 times\n",
      "        34 (100.00%) aligned exactly 1 time\n",
      "        0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "If you use programs that use GNU Parallel to process data for an article in a\n",
      "scientific publication, please cite:\n",
      "\n",
      "  Tange, O. (2022, September 22). GNU Parallel 20220922 ('Elizabeth').\n",
      "  Zenodo. https://doi.org/10.5281/zenodo.7105792\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "More about funding GNU Parallel and the citation notice:\n",
      "https://www.gnu.org/software/parallel/parallel_design.html#citation-notice\n",
      "\n",
      "To silence this citation notice: run 'parallel --citation' once.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# The index name.\n",
    "IDX=refs/genome.fa\n",
    "\n",
    "# Create the BAM folder.\n",
    "mkdir -p bam\n",
    "\n",
    "# Align the FASTQ files to the reference genome.\n",
    "cat ids.txt | parallel \"hisat2 -x $IDX -1 reads/{}_R1.fq -2 reads/{}_R2.fq | samtools sort > bam/{}.bam\"\n",
    "\n",
    "# Index each BAM file.\n",
    "cat ids.txt | parallel  \"samtools index bam/{}.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "00f93677-f3a3-47bb-a8c2-6e9e3e998dc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Academic tradition requires you to cite works you base your article on.\n",
      "If you use programs that use GNU Parallel to process data for an article in a\n",
      "scientific publication, please cite:\n",
      "\n",
      "  Tange, O. (2022, September 22). GNU Parallel 20220922 ('Elizabeth').\n",
      "  Zenodo. https://doi.org/10.5281/zenodo.7105792\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "More about funding GNU Parallel and the citation notice:\n",
      "https://www.gnu.org/software/parallel/parallel_design.html#citation-notice\n",
      "\n",
      "To silence this citation notice: run 'parallel --citation' once.\n",
      "\n",
      "Academic tradition requires you to cite works you base your article on.\n",
      "If you use programs that use GNU Parallel to process data for an article in a\n",
      "scientific publication, please cite:\n",
      "\n",
      "  Tange, O. (2022, September 22). GNU Parallel 20220922 ('Elizabeth').\n",
      "  Zenodo. https://doi.org/10.5281/zenodo.7105792\n",
      "\n",
      "This helps funding further development; AND IT WON'T COST YOU A CENT.\n",
      "If you pay 10000 EUR you should feel free to use GNU Parallel without citing.\n",
      "\n",
      "More about funding GNU Parallel and the citation notice:\n",
      "https://www.gnu.org/software/parallel/parallel_design.html#citation-notice\n",
      "\n",
      "To silence this citation notice: run 'parallel --citation' once.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Turn each BAM file into bedGraph coverage. The files will have the .bg extension.\n",
    "cat ids.txt | parallel \"bedtools genomecov -ibam  bam/{}.bam -split -bg  > bam/{}.bg\"\n",
    "\n",
    "# Convert each bedGraph coverage into bigWig coverage. The files will have the .bw extension.\n",
    "cat ids.txt | parallel \"bedGraphToBigWig bam/{}.bg  ${IDX}.fai bam/{}.bw\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4def30a0-bd33-45aa-ac52-d5d4e8c58fa3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
