{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8ea5e3ee-1d1c-41c6-9eb1-536d40e38b6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 2to3\n",
      " 2to3-3.9\n",
      " acountry\n",
      " activate\n",
      " activate-global-python-argcomplete\n",
      " adig\n",
      " aec\n",
      " ahost\n",
      " alembic\n",
      " aomdec\n",
      " aomenc\n",
      " applygeo\n",
      " aserver\n",
      " assistant\n",
      " attr\n",
      " autopep8\n",
      " autopoint\n",
      " baba\n",
      " balsam\n",
      " bash\n",
      " bashbug\n",
      " bokeh\n",
      " brotli\n",
      " bsdcat\n",
      " bsdcpio\n",
      " bsdtar\n",
      " bunzip2\n",
      " bzcat\n",
      " bzcmp\n",
      " bzdiff\n",
      " bzegrep\n",
      " bzfgrep\n",
      " bzgrep\n",
      " bzip2\n",
      " bzip2recover\n",
      " bzless\n",
      " bzmore\n",
      " cairo-trace\n",
      " canbusutil\n",
      " captoinfo\n",
      " cbrunsli\n",
      " cct\n",
      " certipy\n",
      " certutil\n",
      " cjpeg\n",
      " clear\n",
      " clusterdb\n",
      " compile_et\n",
      " conda\n",
      " conda2solv\n",
      " conda-env\n",
      " corepack\n",
      " cph\n",
      " createdb\n",
      " CreateDOMDocument\n",
      " createuser\n",
      " c_rehash\n",
      " cs2cs\n",
      " cups-config\n",
      " curl\n",
      " curl-config\n",
      " curve_keygen\n",
      " dap-config\n",
      " dap-config-pkgconfig\n",
      " dask\n",
      "'dask scheduler'\n",
      " dask-scheduler\n",
      "'dask ssh'\n",
      " dask-ssh\n",
      "'dask worker'\n",
      " dask-worker\n",
      " dav1d\n",
      " db_archive\n",
      " db_checkpoint\n",
      " db_convert\n",
      " db_deadlock\n",
      " db_dump\n",
      " db_hotbackup\n",
      " db_load\n",
      " db_log_verify\n",
      " db_printlog\n",
      " db_recover\n",
      " db_replicate\n",
      " dbrunsli\n",
      " db_stat\n",
      " db_tuner\n",
      " db_upgrade\n",
      " dbus-cleanup-sockets\n",
      " dbus-daemon\n",
      " dbus-launch\n",
      " dbus-monitor\n",
      " dbus-run-session\n",
      " dbus-send\n",
      " dbus-test-tool\n",
      " dbus-update-activation-environment\n",
      " dbus-uuidgen\n",
      " db_verify\n",
      " deactivate\n",
      " derb\n",
      " designer\n",
      " djpeg\n",
      " dmypy\n",
      " DOMCount\n",
      " DOMPrint\n",
      " dropdb\n",
      " dropuser\n",
      " dumpsolv\n",
      " ecpg\n",
      " EnumVal\n",
      " envsubst\n",
      " epylint\n",
      " esdcompat\n",
      " f2py\n",
      " f2py3\n",
      " f2py3.9\n",
      " fax2ps\n",
      " fax2tiff\n",
      " fc-cache\n",
      " fc-cat\n",
      " fc-conflist\n",
      " fc-list\n",
      " fc-match\n",
      " fc-pattern\n",
      " fc-query\n",
      " fc-scan\n",
      " fc-validate\n",
      " fftwf-wisdom\n",
      " fftwl-wisdom\n",
      " fftw-wisdom\n",
      " fftw-wisdom-to-conf\n",
      " fitscopy\n",
      " fixqt4headers.pl\n",
      " flac\n",
      " flake8\n",
      " fonttools\n",
      " fpack\n",
      " freetype-config\n",
      " fribidi\n",
      " funpack\n",
      " gapplication\n",
      " gdaladdo\n",
      " gdalbuildvrt\n",
      " gdal-config\n",
      " gdal_contour\n",
      " gdal_create\n",
      " gdaldem\n",
      " gdalenhance\n",
      " gdal_grid\n",
      " gdalinfo\n",
      " gdallocationinfo\n",
      " gdalmanage\n",
      " gdalmdiminfo\n",
      " gdalmdimtranslate\n",
      " gdal_rasterize\n",
      " gdalsrsinfo\n",
      " gdaltindex\n",
      " gdaltransform\n",
      " gdal_translate\n",
      " gdal_viewshed\n",
      " gdalwarp\n",
      " gdbus\n",
      " gdbus-codegen\n",
      " genbrk\n",
      " gencfu\n",
      " gencnval\n",
      " gendict\n",
      " genrb\n",
      " geod\n",
      " geos-config\n",
      " geosop\n",
      " geotifcp\n",
      " getdap\n",
      " getdap4\n",
      " getfattr\n",
      " get_objgraph\n",
      " gettext\n",
      " gettextize\n",
      " gettext.sh\n",
      " gie\n",
      " gif2h5\n",
      " gif2hdf\n",
      " gif2rgb\n",
      " gifbuild\n",
      " gifclrmp\n",
      " giffix\n",
      " giftext\n",
      " giftool\n",
      " gio\n",
      " gio-querymodules\n",
      " git2_cli\n",
      " gitpuller\n",
      " glib-compile-resources\n",
      " glib-compile-schemas\n",
      " glib-genmarshal\n",
      " glib-gettextize\n",
      " glib-mkenums\n",
      " gnmanalyse\n",
      " gnmmanage\n",
      " gobject-query\n",
      " gresource\n",
      " gsettings\n",
      " gsl-config\n",
      " gsl-histogram\n",
      " gsl-randist\n",
      " gss-client\n",
      " gst-device-monitor-1.0\n",
      " gst-discoverer-1.0\n",
      " gst-inspect-1.0\n",
      " gst-launch-1.0\n",
      " gst-play-1.0\n",
      " gst-stats-1.0\n",
      " gst-typefind-1.0\n",
      " gtester\n",
      " gtester-report\n",
      " h2load\n",
      " h4cc\n",
      " h4_ncdump\n",
      " h4_ncgen\n",
      " h4redeploy\n",
      " h52gif\n",
      " h5c++\n",
      " h5cc\n",
      " h5clear\n",
      " h5copy\n",
      " h5debug\n",
      " h5diff\n",
      " h5dump\n",
      " h5fc\n",
      " h5format_convert\n",
      " h5import\n",
      " h5jam\n",
      " h5ls\n",
      " h5mkgrp\n",
      " h5perf_serial\n",
      " h5redeploy\n",
      " h5repack\n",
      " h5repart\n",
      " h5stat\n",
      " h5unjam\n",
      " h5watch\n",
      " hb-ot-shape-closure\n",
      " hb-shape\n",
      " hb-subset\n",
      " hb-view\n",
      " hdf24to8\n",
      " hdf2gif\n",
      " hdf2jpeg\n",
      " hdf8to24\n",
      " hdfcomp\n",
      " hdfed\n",
      " hdfimport\n",
      " hdfls\n",
      " hdfpack\n",
      " hdftopal\n",
      " hdftor8\n",
      " hdfunpac\n",
      " hdiff\n",
      " hdp\n",
      " hrepack\n",
      " iconv\n",
      " icu-config\n",
      " icuexportdata\n",
      " icuinfo\n",
      " idle3\n",
      " idle3.9\n",
      " imagecodecs\n",
      " imageio_download_bin\n",
      " imageio_remove_bin\n",
      " imcopy\n",
      " infocmp\n",
      " infotocap\n",
      " initdb\n",
      " installcheck\n",
      " invgeod\n",
      " invproj\n",
      " ippeveprinter\n",
      " ipptool\n",
      " ipython\n",
      " ipython3\n",
      " isort\n",
      " isort-identify-imports\n",
      " jackd\n",
      " jlpm\n",
      " jpeg2hdf\n",
      " jpegtran\n",
      " jpgicc\n",
      " jsonschema\n",
      " jupyter\n",
      " jupyter-bundlerextension\n",
      " jupyter-dejavu\n",
      " jupyter-execute\n",
      " jupyterhub\n",
      " jupyterhub-singleuser\n",
      " jupyter-kernel\n",
      " jupyter-kernelspec\n",
      " jupyter-lab\n",
      " jupyter-labextension\n",
      " jupyter-labhub\n",
      " jupyter-migrate\n",
      " jupyter-nbclassic\n",
      " jupyter-nbconvert\n",
      " jupyter-nbextension\n",
      " jupyter-notebook\n",
      " jupyter-nteract\n",
      " jupyter-retro\n",
      " jupyter-run\n",
      " jupyter-server\n",
      " jupyter-serverextension\n",
      " jupyter-troubleshoot\n",
      " jupyter-trust\n",
      " JxrDecApp\n",
      " JxrEncApp\n",
      " k5srvutil\n",
      " kadmin\n",
      " kdestroy\n",
      " kea-config\n",
      " keyctl\n",
      " kinit\n",
      " klist\n",
      " kpasswd\n",
      " krb5-config\n",
      " ksu\n",
      " kswitch\n",
      " ktutil\n",
      " kvno\n",
      " lame\n",
      " lconvert\n",
      " libdeflate-gunzip\n",
      " libdeflate-gzip\n",
      " libpng16-config\n",
      " libpng-config\n",
      " libtool\n",
      " libtoolize\n",
      " linguist\n",
      " linkicc\n",
      " listgeo\n",
      " lprodump\n",
      " lrelease\n",
      " lrelease-pro\n",
      " lsm2bin\n",
      " lupdate\n",
      " lupdate-pro\n",
      " lz4\n",
      " lz4c\n",
      " lz4cat\n",
      " lzcat\n",
      " lzcmp\n",
      " lzdiff\n",
      " lzegrep\n",
      " lzfgrep\n",
      " lzgrep\n",
      " lzless\n",
      " lzma\n",
      " lzmadec\n",
      " lzmainfo\n",
      " lzmore\n",
      " make\n",
      " makeconv\n",
      " makegeo\n",
      " mako-render\n",
      " mamba\n",
      " mamba-package\n",
      " MemParse\n",
      " mergesolv\n",
      " meshdebug\n",
      " metaflac\n",
      " micromamba\n",
      " moc\n",
      " mpg123\n",
      " mpg123-id3dump\n",
      " mpg123-strip\n",
      " msgattrib\n",
      " msgcat\n",
      " msgcmp\n",
      " msgcomm\n",
      " msgconv\n",
      " msgen\n",
      " msgexec\n",
      " msgfilter\n",
      " msgfmt\n",
      " msggrep\n",
      " msginit\n",
      " msgmerge\n",
      " msgunfmt\n",
      " msguniq\n",
      " mypy\n",
      " mypyc\n",
      " nc-config\n",
      " nccopy\n",
      " ncdump\n",
      " ncgen\n",
      " ncgen3\n",
      " ncurses6-config\n",
      " ncursesw6-config\n",
      " nearblack\n",
      " ngettext\n",
      " nghttp\n",
      " nghttpd\n",
      " nghttpx\n",
      " node\n",
      " normalizer\n",
      " npm\n",
      " npx\n",
      " nspr-config\n",
      " nss-config\n",
      " ogr2ogr\n",
      " ogrinfo\n",
      " ogrlineref\n",
      " ogrtindex\n",
      " oid2name\n",
      " openssl\n",
      " opj_compress\n",
      " opj_decompress\n",
      " opj_dump\n",
      " out123\n",
      " pacat\n",
      " pacmd\n",
      " pactl\n",
      " padsp\n",
      " pa-info\n",
      " pal2rgb\n",
      " paltohdf\n",
      " pamon\n",
      " pandoc\n",
      " pango-list\n",
      " pango-segmentation\n",
      " pango-view\n",
      " paplay\n",
      " parec\n",
      " parecord\n",
      " pasuspender\n",
      " pcre2-config\n",
      " pcre2grep\n",
      " pcre2_jit_test\n",
      " pcre2test\n",
      " pcre-config\n",
      " pcregrep\n",
      " pcretest\n",
      " pdfattach\n",
      " pdfdetach\n",
      " pdffonts\n",
      " pdfimages\n",
      " pdfinfo\n",
      " pdfseparate\n",
      " pdfsig\n",
      " pdftocairo\n",
      " pdftohtml\n",
      " pdftoppm\n",
      " pdftops\n",
      " pdftotext\n",
      " pdfunite\n",
      " pg_amcheck\n",
      " pg_archivecleanup\n",
      " pg_basebackup\n",
      " pgbench\n",
      " pg_checksums\n",
      " pg_config\n",
      " pg_controldata\n",
      " pg_ctl\n",
      " pg_dump\n",
      " pg_dumpall\n",
      " pg_isready\n",
      " pg_receivewal\n",
      " pg_recvlogical\n",
      " pg_resetwal\n",
      " pg_restore\n",
      " pg_rewind\n",
      " pg_test_fsync\n",
      " pg_test_timing\n",
      " pg_upgrade\n",
      " pg_verifybackup\n",
      " pg_waldump\n",
      " pip\n",
      " pip3\n",
      " pipx\n",
      " pixeltool\n",
      " pk12util\n",
      " pkgdata\n",
      " pngfix\n",
      " png-fix-itxt\n",
      " postgres\n",
      " postmaster\n",
      " PParse\n",
      " ppm2tiff\n",
      " proj\n",
      " projinfo\n",
      " projsync\n",
      " psicc\n",
      " psql\n",
      " PSVIWriter\n",
      " pulseaudio\n",
      " pybabel\n",
      " pycodestyle\n",
      " pydoc\n",
      " pydoc3\n",
      " pydoc3.9\n",
      " pydocstyle\n",
      " pyflakes\n",
      " pyftmerge\n",
      " pyftsubset\n",
      " pygmentize\n",
      " pyjson5\n",
      " pylint\n",
      " pylint-config\n",
      " pylsp\n",
      " pylupdate5\n",
      " pyrcc5\n",
      " pyreverse\n",
      " python\n",
      " python3\n",
      " python3.9\n",
      " python3.9-config\n",
      " python3-config\n",
      " python-argcomplete-check-easy-install-script\n",
      " python-argcomplete-tcsh\n",
      " pyuic5\n",
      " qcollectiongenerator\n",
      " qdbus\n",
      " qdbuscpp2xml\n",
      " qdbusviewer\n",
      " qdbusxml2cpp\n",
      " qdistancefieldgenerator\n",
      " qdoc\n",
      " qgltf\n",
      " qhelpgenerator\n",
      " qlalr\n",
      " qmake\n",
      " qml\n",
      " qmlcachegen\n",
      " qmleasing\n",
      " qmlformat\n",
      " qmlimportscanner\n",
      " qmllint\n",
      " qmlmin\n",
      " qmlplugindump\n",
      " qmlpreview\n",
      " qmlprofiler\n",
      " qmlscene\n",
      " qmltestrunner\n",
      " qmltime\n",
      " qmltyperegistrar\n",
      " qpaeq\n",
      " qscxmlc\n",
      " qtattributionsscanner\n",
      " qt.conf\n",
      " qtdiag\n",
      " qtpaths\n",
      " qtplugininfo\n",
      " qvkgen\n",
      " R\n",
      " r8tohdf\n",
      " raw2tiff\n",
      " rcc\n",
      " rdjpgcom\n",
      " recode-sr-latin\n",
      " Redirect\n",
      " register-python-argcomplete\n",
      " reindexdb\n",
      " repc\n",
      " repo2solv\n",
      " reset\n",
      " ristosds\n",
      " Rscript\n",
      " SAX2Count\n",
      " SAX2Print\n",
      " SAXCount\n",
      " SAXPrint\n",
      " sclient\n",
      " SCMPrint\n",
      " sed\n",
      " send2trash\n",
      " SEnumVal\n",
      " setfattr\n",
      " sim_client\n",
      " sip-build\n",
      " sip-distinfo\n",
      " sip-install\n",
      " sip-module\n",
      " sip-sdist\n",
      " sip-wheel\n",
      " skivi\n",
      " smem\n",
      " sndfile-cmp\n",
      " sndfile-concat\n",
      " sndfile-convert\n",
      " sndfile-deinterleave\n",
      " sndfile-info\n",
      " sndfile-interleave\n",
      " sndfile-metadata-get\n",
      " sndfile-metadata-set\n",
      " sndfile-play\n",
      " sndfile-salvage\n",
      " sqlite3\n",
      " sqlite3_analyzer\n",
      " StdInParse\n",
      " stubgen\n",
      " stubtest\n",
      " symilar\n",
      " syncqt.pl\n",
      " tabs\n",
      " tclsh\n",
      " tclsh8.6\n",
      " testsolv\n",
      " tic\n",
      " tiff2bw\n",
      " tiff2fsspec\n",
      " tiff2pdf\n",
      " tiff2ps\n",
      " tiff2rgba\n",
      " tiffcmp\n",
      " tiffcomment\n",
      " tiffcp\n",
      " tiffcrop\n",
      " tiffdither\n",
      " tiffdump\n",
      " tifffile\n",
      " tiffinfo\n",
      " tiffmedian\n",
      " tiffset\n",
      " tiffsplit\n",
      " tificc\n",
      " toe\n",
      " tput\n",
      " tqdm\n",
      " tracegen\n",
      " transicc\n",
      " tset\n",
      " ttx\n",
      " tzselect\n",
      " uic\n",
      " undill\n",
      " unlz4\n",
      " unlzma\n",
      " unxz\n",
      " unzstd\n",
      " userpath\n",
      " uuclient\n",
      " vacuumdb\n",
      " vacuumlo\n",
      " vmake\n",
      " vshow\n",
      " wheel\n",
      " wish\n",
      " wish8.6\n",
      " wrjpgcom\n",
      " wsdump\n",
      " x86_64-conda_cos6-linux-gnu-addr2line\n",
      " x86_64-conda_cos6-linux-gnu-ar\n",
      " x86_64-conda_cos6-linux-gnu-as\n",
      " x86_64-conda_cos6-linux-gnu-c++\n",
      " x86_64-conda_cos6-linux-gnu-c89\n",
      " x86_64-conda_cos6-linux-gnu-c99\n",
      " x86_64-conda_cos6-linux-gnu-cc\n",
      " x86_64-conda_cos6-linux-gnu-c++filt\n",
      " x86_64-conda_cos6-linux-gnu-cpp\n",
      " x86_64-conda_cos6-linux-gnu-dwp\n",
      " x86_64-conda_cos6-linux-gnu-elfedit\n",
      " x86_64-conda_cos6-linux-gnu-f95\n",
      " x86_64-conda_cos6-linux-gnu-g++\n",
      " x86_64-conda_cos6-linux-gnu-gcc\n",
      " x86_64-conda_cos6-linux-gnu-gcc-ar\n",
      " x86_64-conda_cos6-linux-gnu-gcc-nm\n",
      " x86_64-conda_cos6-linux-gnu-gcc-ranlib\n",
      " x86_64-conda_cos6-linux-gnu-gcov\n",
      " x86_64-conda_cos6-linux-gnu-gcov-dump\n",
      " x86_64-conda_cos6-linux-gnu-gcov-tool\n",
      " x86_64-conda_cos6-linux-gnu-gfortran\n",
      " x86_64-conda_cos6-linux-gnu-gprof\n",
      " x86_64-conda_cos6-linux-gnu-ld\n",
      " x86_64-conda_cos6-linux-gnu-ld.bfd\n",
      " x86_64-conda_cos6-linux-gnu-ld.gold\n",
      " x86_64-conda_cos6-linux-gnu-nm\n",
      " x86_64-conda_cos6-linux-gnu-objcopy\n",
      " x86_64-conda_cos6-linux-gnu-objdump\n",
      " x86_64-conda_cos6-linux-gnu-ranlib\n",
      " x86_64-conda_cos6-linux-gnu-readelf\n",
      " x86_64-conda_cos6-linux-gnu-size\n",
      " x86_64-conda_cos6-linux-gnu-strings\n",
      " x86_64-conda_cos6-linux-gnu-strip\n",
      " x86_64-conda-linux-gnu-addr2line\n",
      " x86_64-conda-linux-gnu-ar\n",
      " x86_64-conda-linux-gnu-as\n",
      " x86_64-conda-linux-gnu-c++\n",
      " x86_64-conda-linux-gnu-c89\n",
      " x86_64-conda-linux-gnu-c99\n",
      " x86_64-conda-linux-gnu-cc\n",
      " x86_64-conda-linux-gnu-c++filt\n",
      " x86_64-conda-linux-gnu-cpp\n",
      " x86_64-conda-linux-gnu-dwp\n",
      " x86_64-conda-linux-gnu-elfedit\n",
      " x86_64-conda-linux-gnu-f95\n",
      " x86_64-conda-linux-gnu-g++\n",
      " x86_64-conda-linux-gnu-gcc\n",
      " x86_64-conda-linux-gnu-gcc-ar\n",
      " x86_64-conda-linux-gnu-gcc-nm\n",
      " x86_64-conda-linux-gnu-gcc-ranlib\n",
      " x86_64-conda-linux-gnu-gcov\n",
      " x86_64-conda-linux-gnu-gcov-dump\n",
      " x86_64-conda-linux-gnu-gcov-tool\n",
      " x86_64-conda-linux-gnu-gfortran\n",
      " x86_64-conda-linux-gnu-gprof\n",
      " x86_64-conda-linux-gnu-ld\n",
      " x86_64-conda-linux-gnu-ld.bfd\n",
      " x86_64-conda-linux-gnu-ld.gold\n",
      " x86_64-conda-linux-gnu-nm\n",
      " x86_64-conda-linux-gnu-objcopy\n",
      " x86_64-conda-linux-gnu-objdump\n",
      " x86_64-conda-linux-gnu-ranlib\n",
      " x86_64-conda-linux-gnu-readelf\n",
      " x86_64-conda-linux-gnu-size\n",
      " x86_64-conda-linux-gnu-strings\n",
      " x86_64-conda-linux-gnu-strip\n",
      " xgettext\n",
      " XInclude\n",
      " xkbcli\n",
      " xml2-config\n",
      " xmlcatalog\n",
      " xmllint\n",
      " xmlpatterns\n",
      " xmlpatternsvalidator\n",
      " xmlwf\n",
      " xz\n",
      " xzcat\n",
      " xzcmp\n",
      " xzdec\n",
      " xzdiff\n",
      " xzegrep\n",
      " xzfgrep\n",
      " xzgrep\n",
      " xzless\n",
      " xzmore\n",
      " yapf\n",
      " yapf-diff\n",
      " zdump\n",
      " zfp\n",
      " zic\n",
      " zipcmp\n",
      " zipmerge\n",
      " ziptool\n",
      " zopfli\n",
      " zopflipng\n",
      " zstd\n",
      " zstdcat\n",
      " zstdgrep\n",
      " zstdless\n",
      " zstdmt\n"
     ]
    }
   ],
   "source": [
    "pand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d47372a1-1277-4358-98a3-c7042a09ed78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Info: This is the most recent version of salmon.\n",
      "index [\"refs/salmon.idx\"] did not previously exist  . . . creating it\n",
      "[2022-10-20 03:15:33.861] [jLog] [warning] The salmon index is being built without any decoy sequences.  It is recommended that decoy sequence (either computed auxiliary decoy sequence or the genome of the organism) be provided during indexing. Further details can be found at https://salmon.readthedocs.io/en/latest/salmon.html#preparing-transcriptome-indices-mapping-based-mode.\n",
      "[2022-10-20 03:15:33.861] [jLog] [info] building index\n",
      "out : refs/salmon.idx\n",
      "[2022-10-20 03:15:33.861] [puff::index::jointLog] [info] Running fixFasta\n",
      "\n",
      "[Step 1 of 4] : counting k-mers\n",
      "\n",
      "[2022-10-20 03:15:33.875] [puff::index::jointLog] [info] Replaced 1 non-ATCG nucleotides\n",
      "[2022-10-20 03:15:33.875] [puff::index::jointLog] [info] Clipped poly-A tails from 91 transcripts\n",
      "wrote 92 cleaned references\n",
      "[2022-10-20 03:15:33.878] [puff::index::jointLog] [info] Filter size not provided; estimating from number of distinct k-mers\n",
      "[2022-10-20 03:15:33.881] [puff::index::jointLog] [info] ntHll estimated 84274 distinct k-mers, setting filter size to 2^21\n",
      "Threads = 2\n",
      "Vertex length = 31\n",
      "Hash functions = 5\n",
      "Filter size = 2097152\n",
      "Capacity = 2\n",
      "Files: \n",
      "refs/salmon.idx/ref_k31_fixed.fa\n",
      "--------------------------------------------------------------------------------\n",
      "Round 0, 0:2097152\n",
      "Pass\tFilling\tFiltering\n",
      "1\t0\t0\t\n",
      "2\t0\t0\n",
      "True junctions count = 0\n",
      "False junctions count = 318\n",
      "Hash table size = 318\n",
      "Candidate marks count = 318\n",
      "--------------------------------------------------------------------------------\n",
      "Reallocating bifurcations time: 0\n",
      "True marks count: 184\n",
      "Edges construction time: 0\n",
      "--------------------------------------------------------------------------------\n",
      "Distinct junctions = 0\n",
      "\n",
      "TwoPaCo::buildGraphMain:: allocated with scalable_malloc; freeing.\n",
      "TwoPaCo::buildGraphMain:: Calling scalable_allocation_command(TBBMALLOC_CLEAN_ALL_BUFFERS, 0);\n",
      "allowedIn: 21\n",
      "Max Junction ID: 225\n",
      "seen.size():1809 kmerInfo.size():226\n",
      "approximateContigTotalLength: 77736\n",
      "counters for complex kmers:\n",
      "(prec>1 & succ>1)=0 | (succ>1 & isStart)=0 | (prec>1 & isEnd)=0 | (isStart & isEnd)=0\n",
      "contig count: 93 element count: 80588 complex nodes: 0\n",
      "# of ones in rank vector: 92\n",
      "[2022-10-20 03:15:33.995] [puff::index::jointLog] [info] Starting the Pufferfish indexing by reading the GFA binary file.\n",
      "[2022-10-20 03:15:33.995] [puff::index::jointLog] [info] Setting the index/BinaryGfa directory refs/salmon.idx\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading contigs | Time = 1.7835 ms\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 56.76 us\n",
      "-----------------------------------------\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "92\n",
      "[2022-10-20 03:15:33.997] [puff::index::jointLog] [info] Done wrapping the rank vector with a rank9sel structure.\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] contig count for validation: 92\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] Total # of Contigs : 92\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] Total # of numerical Contigs : 92\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] Total # of contig vec entries: 92\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] bits per offset entry 7\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] Done constructing the contig vector. 93\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] # segments = 92\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] total length = 80,588\n",
      "[2022-10-20 03:15:34.004] [puff::index::jointLog] [info] Reading the reference files ...\n",
      "[2022-10-20 03:15:34.007] [puff::index::jointLog] [info] positional integer width = 17\n",
      "[2022-10-20 03:15:34.007] [puff::index::jointLog] [info] seqSize = 80,588\n",
      "[2022-10-20 03:15:34.007] [puff::index::jointLog] [info] rankSize = 80,588\n",
      "[2022-10-20 03:15:34.007] [puff::index::jointLog] [info] edgeVecSize = 0\n",
      "[2022-10-20 03:15:34.007] [puff::index::jointLog] [info] num keys = 77,828\n",
      "for info, total work write each  : 2.331    total work inram from level 3 : 4.322  total work raw : 25.000 \n",
      "[Building BooPHF]  100  %   elapsed:   0 min 0  sec   remaining:   0 min 0  sec\n",
      "Bitarray          413824  bits (100.00 %)   (array + ranks )\n",
      "final hash             0  bits (0.00 %) (nb in final hash 0)\n",
      "[2022-10-20 03:15:34.063] [puff::index::jointLog] [info] mphf size = 0.0493317 MB\n",
      "[2022-10-20 03:15:34.063] [puff::index::jointLog] [info] chunk size = 40,294\n",
      "[2022-10-20 03:15:34.063] [puff::index::jointLog] [info] chunk 0 = [0, 40,294)\n",
      "[2022-10-20 03:15:34.063] [puff::index::jointLog] [info] chunk 1 = [40,294, 80,558)\n",
      "[2022-10-20 03:15:34.071] [puff::index::jointLog] [info] finished populating pos vector\n",
      "[2022-10-20 03:15:34.071] [puff::index::jointLog] [info] writing index components\n",
      "[2022-10-20 03:15:34.072] [puff::index::jointLog] [info] finished writing dense pufferfish index\n",
      "[2022-10-20 03:15:34.072] [jLog] [info] done building index\n"
     ]
    }
   ],
   "source": [
    "# The transcriptome in fasta format.\n",
    "REF=refs/transcripts.fa\n",
    "\n",
    "# The name of the salmon index\n",
    "IDX=refs/salmon.idx\n",
    "\n",
    "# Build the index with salmon.\n",
    "salmon index -t ${REF} -i ${IDX}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be46f914-9372-42ba-a84b-8e4e0e4739d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/BORED_1_R1.fq }\n",
      "### [ mates2 ] => { reads/BORED_1_R2.fq }\n",
      "### [ output ] => { salmon/BORED_1 }\n",
      "Logs will be written to salmon/BORED_1/logs\n",
      "[2022-10-20 03:16:18.954] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:16:18.954] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:16:18.954] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:16:18.954] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:16:18.954] [jointLog] [info] parsing read library format\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 38.935 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 59.408 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 6.193 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 284.53 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 4.582 ms\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 167.94 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 3.0933 ms\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 72.446 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 9.789 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:16:18.956] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:16:18.956] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:16:18.956] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:16:18.965] [jointLog] [info] done\n",
      "[2022-10-20 03:16:19.092] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:16:19.094] [jointLog] [info] Number of decoys : 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:16:19.357] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "[2022-10-20 03:16:19.537] [jointLog] [info] Thread saw mini-batch with a maximum of 0.06% zero probability fragments\n",
      "[2022-10-20 03:16:19.539] [jointLog] [info] Thread saw mini-batch with a maximum of 0.04% zero probability fragments\n",
      "[2022-10-20 03:16:19.551] [jointLog] [info] Computed 66 rich equivalence classes for further processing\n",
      "[2022-10-20 03:16:19.551] [jointLog] [info] Counted 112,173 total reads in the equivalence classes \n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] Number of mappings discarded because of alignment score : 6\n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 6\n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 704\n",
      "[2022-10-20 03:16:19.553] [jointLog] [warning] Only 112173 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] Mapping rate = 99.9822%\n",
      "\n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:16:19.553] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:16:19.554] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:16:19.554] [jointLog] [info] iteration = 0 | max rel diff. = 1215.54\n",
      "[2022-10-20 03:16:19.556] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:16:19.556] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:16:19.556] [jointLog] [info] writing output \n",
      "\n"
     ]
    }
   ],
   "source": [
    "mkdir -p salmon\n",
    "\n",
    "salmon quant -i \"$IDX\" -l A -1 reads/BORED_1_R1.fq -2 reads/BORED_1_R2.fq  -o salmon/BORED_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "336053a1-1c99-49d0-908a-4ccc0c4ede75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/BORED_3_R1.fq }\n",
      "### [ mates2 ] => { reads/BORED_3_R2.fq }\n",
      "### [ output ] => { salmon/BORED_3 }\n",
      "Logs will be written to salmon/BORED_3/logs\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 34.631 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 41.091 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 3.486 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 102.89 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 208.17 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 52.964 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 154.12 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 50.364 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 6.69 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] parsing read library format\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:17:21.028] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:17:21.029] [jointLog] [info] done\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:21.417] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:17:21.417] [jointLog] [info] Number of decoys : 0\n",
      "[2022-10-20 03:17:22.614] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:23.440] [jointLog] [info] Thread saw mini-batch with a maximum of 0.04% zero probability fragments\n",
      "[2022-10-20 03:17:23.458] [jointLog] [info] Thread saw mini-batch with a maximum of 0.08% zero probability fragments\n",
      "[2022-10-20 03:17:23.475] [jointLog] [info] Computed 67 rich equivalence classes for further processing\n",
      "[2022-10-20 03:17:23.475] [jointLog] [info] Counted 123,065 total reads in the equivalence classes \n",
      "[2022-10-20 03:17:23.476] [jointLog] [info] Number of mappings discarded because of alignment score : 10\n",
      "[2022-10-20 03:17:23.476] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 10\n",
      "[2022-10-20 03:17:23.476] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:17:23.476] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 824\n",
      "[2022-10-20 03:17:23.476] [jointLog] [warning] Only 123065 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:17:23.476] [jointLog] [info] Mapping rate = 99.9773%\n",
      "\n",
      "[2022-10-20 03:17:23.477] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:17:23.477] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:17:23.478] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:17:23.478] [jointLog] [info] iteration = 0 | max rel diff. = 1333.37\n",
      "[2022-10-20 03:17:23.480] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:17:23.480] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:17:23.480] [jointLog] [info] writing output \n",
      "\n",
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/BORED_1_R1.fq }\n",
      "### [ mates2 ] => { reads/BORED_1_R2.fq }\n",
      "### [ output ] => { salmon/BORED_1 }\n",
      "Logs will be written to salmon/BORED_1/logs\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 40.008 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 46.08 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 3.454 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 106.24 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 198.57 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 55.389 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 158.26 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 51.399 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 6.74 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] parsing read library format\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:17:21.024] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:17:21.025] [jointLog] [info] done\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:21.450] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:17:21.450] [jointLog] [info] Number of decoys : 0\n",
      "[2022-10-20 03:17:22.556] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:23.194] [jointLog] [info] Thread saw mini-batch with a maximum of 0.06% zero probability fragments\n",
      "[2022-10-20 03:17:23.247] [jointLog] [info] Thread saw mini-batch with a maximum of 0.04% zero probability fragments\n",
      "[2022-10-20 03:17:23.273] [jointLog] [info] Computed 66 rich equivalence classes for further processing\n",
      "[2022-10-20 03:17:23.273] [jointLog] [info] Counted 112,172 total reads in the equivalence classes \n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] Number of mappings discarded because of alignment score : 6\n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 6\n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 704\n",
      "[2022-10-20 03:17:23.281] [jointLog] [warning] Only 112172 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] Mapping rate = 99.9813%\n",
      "\n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:17:23.281] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:17:23.381] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:17:23.381] [jointLog] [info] iteration = 0 | max rel diff. = 1215.53\n",
      "[2022-10-20 03:17:23.383] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:17:23.383] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:17:23.383] [jointLog] [info] writing output \n",
      "\n",
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/BORED_2_R1.fq }\n",
      "### [ mates2 ] => { reads/BORED_2_R2.fq }\n",
      "### [ output ] => { salmon/BORED_2 }\n",
      "Logs will be written to salmon/BORED_2/logs\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 34.498 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 40.575 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 3.221 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 93.545 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 195.25 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 51.208 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 148.24 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 48.852 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 6.027 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] parsing read library format\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:17:21.031] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] done\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:21.464] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:17:21.464] [jointLog] [info] Number of decoys : 0\n",
      "[2022-10-20 03:17:22.506] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:23.546] [jointLog] [info] Thread saw mini-batch with a maximum of 0.04% zero probability fragments\n",
      "[2022-10-20 03:17:23.557] [jointLog] [info] Thread saw mini-batch with a maximum of 0.06% zero probability fragments\n",
      "[2022-10-20 03:17:23.572] [jointLog] [info] Computed 1,249 rich equivalence classes for further processing\n",
      "[2022-10-20 03:17:23.572] [jointLog] [info] Counted 137,565 total reads in the equivalence classes \n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] Number of mappings discarded because of alignment score : 5\n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 5\n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 4,288\n",
      "[2022-10-20 03:17:23.574] [jointLog] [warning] Only 137565 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] Mapping rate = 99.9884%\n",
      "\n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:17:23.574] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:17:23.575] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:17:23.575] [jointLog] [info] iteration = 0 | max rel diff. = 2385.57\n",
      "[2022-10-20 03:17:23.580] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:17:23.580] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:17:23.580] [jointLog] [info] writing output \n",
      "\n",
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/EXCITED_1_R1.fq }\n",
      "### [ mates2 ] => { reads/EXCITED_1_R2.fq }\n",
      "### [ output ] => { salmon/EXCITED_1 }\n",
      "Logs will be written to salmon/EXCITED_1/logs\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 35.905 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 38.213 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 3.314 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 94.089 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 195.89 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 50.023 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 141.3 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 48.101 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 6.135 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] parsing read library format\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:17:21.032] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:17:21.033] [jointLog] [info] done\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:21.395] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:17:21.396] [jointLog] [info] Number of decoys : 0\n",
      "[2022-10-20 03:17:22.662] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:24.513] [jointLog] [info] Thread saw mini-batch with a maximum of 0.12% zero probability fragments\n",
      "[2022-10-20 03:17:24.524] [jointLog] [info] Thread saw mini-batch with a maximum of 0.10% zero probability fragments\n",
      "[2022-10-20 03:17:24.543] [jointLog] [info] Computed 74 rich equivalence classes for further processing\n",
      "[2022-10-20 03:17:24.543] [jointLog] [info] Counted 236,933 total reads in the equivalence classes \n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] Number of mappings discarded because of alignment score : 8\n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 8\n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 730\n",
      "[2022-10-20 03:17:24.544] [jointLog] [warning] Only 236933 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] Mapping rate = 99.9641%\n",
      "\n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:17:24.544] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:17:24.553] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:17:24.553] [jointLog] [info] iteration = 0 | max rel diff. = 2562.16\n",
      "[2022-10-20 03:17:24.555] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:17:24.555] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:17:24.555] [jointLog] [info] writing output \n",
      "\n",
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/EXCITED_2_R1.fq }\n",
      "### [ mates2 ] => { reads/EXCITED_2_R2.fq }\n",
      "### [ output ] => { salmon/EXCITED_2 }\n",
      "Logs will be written to salmon/EXCITED_2/logs\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 37.879 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 45.507 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 3.445 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 235.31 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 203.43 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 49.718 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 155.24 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 47.185 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 6.191 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] parsing read library format\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:17:25.248] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:17:25.249] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:17:25.250] [jointLog] [info] done\n",
      "[2022-10-20 03:17:25.339] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:17:25.339] [jointLog] [info] Number of decoys : 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:25.832] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:26.769] [jointLog] [info] Thread saw mini-batch with a maximum of 0.12% zero probability fragments\n",
      "[2022-10-20 03:17:26.782] [jointLog] [info] Thread saw mini-batch with a maximum of 0.10% zero probability fragments\n",
      "[2022-10-20 03:17:26.798] [jointLog] [info] Computed 72 rich equivalence classes for further processing\n",
      "[2022-10-20 03:17:26.798] [jointLog] [info] Counted 157,959 total reads in the equivalence classes \n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] Number of mappings discarded because of alignment score : 12\n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 12\n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 1,109\n",
      "[2022-10-20 03:17:26.799] [jointLog] [warning] Only 157959 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] Mapping rate = 99.9684%\n",
      "\n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:17:26.799] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:17:26.800] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:17:26.800] [jointLog] [info] iteration = 0 | max rel diff. = 1710.52\n",
      "[2022-10-20 03:17:26.802] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:17:26.802] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:17:26.802] [jointLog] [info] writing output \n",
      "\n",
      "Version Info: This is the most recent version of salmon.\n",
      "### salmon (selective-alignment-based) v1.9.0\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { refs/salmon.idx }\n",
      "### [ libType ] => { A }\n",
      "### [ validateMappings ] => { }\n",
      "### [ mates1 ] => { reads/EXCITED_3_R1.fq }\n",
      "### [ mates2 ] => { reads/EXCITED_3_R2.fq }\n",
      "### [ output ] => { salmon/EXCITED_3 }\n",
      "Logs will be written to salmon/EXCITED_3/logs\n",
      "-----------------------------------------\n",
      "| Loading contig table | Time = 37.457 us\n",
      "-----------------------------------------\n",
      "size = 93\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 41.415 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 3.34 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 102.99 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "Number of ones: 92\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 1\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 200.26 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 49.727 us\n",
      "-----------------------------------------\n",
      "size = 77828\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 146.98 us\n",
      "-----------------------------------------\n",
      "size = 80588\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 53.471 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 6.695 us\n",
      "-----------------------------------------\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] setting maxHashResizeThreads to 2\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] parsing read library format\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] There is 1 library.\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] Loading pufferfish index\n",
      "[2022-10-20 03:17:25.251] [jointLog] [info] Loading dense pufferfish index.\n",
      "[2022-10-20 03:17:25.252] [jointLog] [info] done\n",
      "[2022-10-20 03:17:25.340] [jointLog] [info] Index contained 92 targets\n",
      "[2022-10-20 03:17:25.340] [jointLog] [info] Number of decoys : 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:25.856] [jointLog] [info] Automatically detected most likely library type as ISR\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[2022-10-20 03:17:27.111] [jointLog] [info] Thread saw mini-batch with a maximum of 0.12% zero probability fragments\n",
      "[2022-10-20 03:17:27.121] [jointLog] [info] Thread saw mini-batch with a maximum of 0.12% zero probability fragments\n",
      "[2022-10-20 03:17:27.135] [jointLog] [info] Computed 72 rich equivalence classes for further processing\n",
      "[2022-10-20 03:17:27.135] [jointLog] [info] Counted 196,593 total reads in the equivalence classes \n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] Number of mappings discarded because of alignment score : 20\n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] Number of fragments entirely discarded because of alignment score : 20\n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] Number of fragments discarded because they are best-mapped to decoys : 0\n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] Number of fragments discarded because they have only dovetail (discordant) mappings to valid targets : 1,264\n",
      "[2022-10-20 03:17:27.137] [jointLog] [warning] Only 196593 fragments were mapped, but the number of burn-in fragments was set to 5000000.\n",
      "The effective lengths have been computed using the observed mappings.\n",
      "\n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] Mapping rate = 99.9593%\n",
      "\n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] finished quantifyLibrary()\n",
      "[2022-10-20 03:17:27.137] [jointLog] [info] Starting optimizer\n",
      "[2022-10-20 03:17:27.138] [jointLog] [info] Marked 0 weighted equivalence classes as degenerate\n",
      "[2022-10-20 03:17:27.138] [jointLog] [info] iteration = 0 | max rel diff. = 2127.48\n",
      "[2022-10-20 03:17:27.139] [jointLog] [info] iteration = 100 | max rel diff. = 0\n",
      "[2022-10-20 03:17:27.139] [jointLog] [info] Finished optimizer\n",
      "[2022-10-20 03:17:27.139] [jointLog] [info] writing output \n",
      "\n"
     ]
    }
   ],
   "source": [
    "cat ids.txt | parallel -j 4 \"salmon quant -i ${IDX} -l A -1 reads/{}_R1.fq -2 reads/{}_R2.fq  -o salmon/{}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8541d46c-a2a5-438d-bfe8-9154d50f5b8d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
